Fixes #11035 - template preview feature

ares opened the pull request:
It's small PR, the biggest part is moving variables preparation from unattended controller to generic Renderer module.

dLobatog commented:
[test]

ohadlevy commented on app/models/hostgroup.rb:
```diff
@@ -106,7 +106,7 @@ def hostgroup
 
   def diskLayout
-    ptable.layout.gsub("\r","")
+    ptable.layout.tr("\r", '')
```
switching to tr because its faster?

ares replied:
:dart:

ohadlevy commented on app/views/templates/_form.html.erb:
```diff
@@ -43,16 +43,26 @@
+              <input type="radio" name="options" id="option3" ><%= _("Preview") %>
             </label>
           </div>
+          <span style="display:none" id="preview_host_selector">
```
you can use css class hide instead of display:none

ohadlevy commented on app/views/templates/_form.html.erb:
```diff
@@ -43,16 +43,26 @@
         </div>
       </div>
-      <%= textarea_f f, :template, :class => "template_text", :label =>:none, :disabled => @template.locked?, :size => "max", :'data-file-name' => @template.name %>
+      <div id="preview_error" class="alert alert-danger alert-dismissable" style="display:none">
```
consider using the alert helpers?

ares replied:
I had to slightly enhance alert helper, converted

ohadlevy commented on config/routes.rb:
```diff
@@ -266,6 +266,7 @@
           get 'clone_template'
           get 'lock'
           get 'unlock'
+          post 'preview'
```
does it make any sense to have a similar action over the api?

ares replied:
I'd like to keep this PR minimal so it can still get into 1.9, adding API could be interesting but probably not that useful as in UI where you can modify template and preview it even before saving (with API you'd have to specify the whole template for preview and then do second query to save it)

ohadlevy commented:
tests?

ares commented:
> tests?

added functional tests for ptables and provisioning templates, not sure what else to test, I think I addressed all your comments, could you please re-review?

domcleal commented:
There are currently some test failures.

ares committed:
Fixes #11035 - template preview feature

domcleal commented on app/controllers/templates_controller.rb:
```diff
@@ -78,8 +78,26 @@ def auto_complete_controller_name
+      load_template_vars
+      render :text => unattended_render(template) and return
+    rescue => error
+      msg = _("There was an error rendering the %s template: ") % (template.name)
```
I think I saw we have this format elsewhere, but I'd suggest avoiding translating messages that are logged, and to include the %s for the error.message in the translated string - don't just concatenate.

ares replied:
I think I've fixed the first part, not sure how would you like me to fix the second one. Could you please show me the direction?

domcleal replied:
<pre>
Foreman::Logging.exception("Error rendering the #{template.name} template", error)
render :text => _("There was an error rendering the %{name} template: %{error}") % {:name => template.name, :error => error.message}, :status => :internal_server_error
</pre>


I think you'll just have to use two strings, since they're used differently.

domcleal commented on app/controllers/templates_controller.rb:
```diff
@@ -78,8 +78,26 @@ def auto_complete_controller_name
+  def safe_render(template)
+    begin
+      load_template_vars
+      render :text => unattended_render(template) and return
```
Any need for `and return` on this and the line below?

domcleal commented on app/views/templates/_form.html.erb:
```diff
@@ -43,16 +43,25 @@
             </label>
           </div>
+          <span id="preview_host_selector" style="display:none">
+            <%= select_tag :preview_host_id, options_from_collection_for_select(Host.authorized(:view_hosts).limit(100), :id, :name),
```
This list seems to be unsorted, it should probably be alphabetised.

domcleal replied:
Perhaps also search for hosts that are running an associated OS, so the list is more accurate

domcleal replied:
And managed hosts only?

ares replied:
Good points! Order makes sense, managed hosts make sense for Ptable and ProvisioningTemplate, leaving without it for other potential templates. Associated OSes does not seem to work nicely here, note that until we have the template saved, we don't have the associations so we couldn't display any host - usually the case for new templates or if there are no host of such OS yet. Snippets don't even have associations (but e.g. networking one is OS specific). Also previewing templates might be useful for host with other OS just to see if it works.

domcleal replied:
Makes sense, thanks

domcleal commented on app/assets/javascripts/provisioning_template.js:
```diff
@@ -136,8 +154,35 @@ function set_diff_mode(item){
+  }).fail(function(response){
+    $("div#preview_error span.text").html(response.responseText);
+    $("div#preview_error").show();
+    session.setValue(__('There was an error during rendering, fix it in code and retry'));
```
"There was an error during rendering, return to the Code tab to edit the template."?  ("Fix it in code" sounds odd.)

domcleal commented on lib/foreman/renderer.rb:
```diff
@@ -4,7 +4,6 @@ module Foreman
   module Renderer
     ALLOWED_HELPERS = [ :foreman_url, :grub_pass, :snippet, :snippets,
                         :snippet_if_exists, :ks_console, :root_pass,
-                        :multiboot, :jumpstart_path, :install_path, :miniroot,
```
Is this part of #2525 instead?

ares replied:
your global overview keeps to amaze me

domcleal commented:
Bootdisk adds its own template kind, which can't be rendered through the unattended mechanism, it has its own renderer.  I'd quite like a way to either - disable rendering for certain template kinds (or vice versa), or to supply my own renderer or something.  Any ideas?

domcleal commented on app/models/host/managed.rb:
```diff
@@ -284,7 +284,8 @@ def handle_realm
   def diskLayout
     @host = self
-    pxe_render((disk.empty? ? ptable.layout : disk).gsub("\r",""))
+    template = disk.blank? ? ptable.layout : disk
```
Long-standing bug right on this line is that if `ptable.nil?` then you get: `»There was an error rendering the Kickstart default template: «undefined method 'layout' for nil:NilClass`, perhaps you'd be kind enough to fix it?

ares replied:
fixed via https://github.com/theforeman/foreman/pull/2524/files#diff-6632a0077c474a44423343eb597f9d72R158, I thought raising exception is actually better than rendering empty string as ptable which would cause an issue too late

EDIT: note that there are extra ifs (IIRC on two lines)

domcleal replied:
Looks good.

ares commented:
> Bootdisk adds its own template kind, which can't be rendered through the unattended mechanism, it has its own renderer. I'd quite like a way to either - disable rendering for certain template kinds (or vice versa), or to supply my own renderer or something. Any ideas?

I'd suggest considering converting these templates to it's own subclass and implement their own controller (inheriting from TemplatesController), and overriding the preview action, similar to what we do in remote execution plugin for JobTemplates https://github.com/ares/foreman_remote_execution/commit/643407d55ab439dcf88541aed0cd2818036a4e17

If it's too complicated (API changes etc) we could make the renderer injectable but I'm not sure about entrypoint you could use from bootdisk. If template kind is the only way how to differentiate bootdisk templates, we don't have much other options. Disabling the preview for bootdisk templates would be the easiest solution but quite sad at the same time.

domcleal commented:
> Bootdisk adds its own template kind, which can't be rendered through the unattended mechanism, it has its own renderer. I'd quite like a way to either - disable rendering for certain template kinds (or vice versa), or to supply my own renderer or something. Any ideas?
> 
> I'd suggest considering converting these templates to it's own subclass and implement their own controller (inheriting from TemplatesController), and overriding the preview action

Yeah, I agree.  I was thinking about it and realised that they're not actually unattended provisioning templates, so relying on the behaviour of using the unattended renderer doesn't make sense.  Thanks, I'll fix this there some time.

ares commented:
@domcleal I think I fixed all comments, would you mind giving it another look? Could this get to 1.9 branch? I think it's quite isolated (the worst case is that bootdisk templates preview won't work I guess). Reasons why I think it would be beneficial - it helps debugging templates, fixes the rendering issue when reviewing templates of host that were provisioned using image, foreman_remote_execution would greatly benefit from it.

domcleal commented:
Hm, I'll perhaps set it to 1.9.1 and see how it settles in in the meantime, with regressions etc.  I agree it's a help, but it would usually go to the next release.

There's a bug when no hosts are available that match the criteria - the AJAX passes `"preview_host_id"=>"null"` and then you get a large 404 error in the page:

![screenshot from 2015-07-15 10 40 01](https://cloud.githubusercontent.com/assets/482089/8695642/33760f5c-2ade-11e5-8f9d-25b811227a19.png)

ares commented:
> Hm, I'll perhaps set it to 1.9.1 and see how it settles in in the meantime, with regressions etc. I agree it's a help, but it would usually go to the next release.

I was hoping for this to be a candidate for "Small and low risk ones may even make it through to 1.9." stated in branch out announcement :-) 1.9.1 would be probably too late for remote execution. What if I promise fixing any related issue, we still have at least one RC to test. If it's no-go we'll have to find some way how to extend Foreman core JS (hints?) from plugin and convert everything to concerns.

> There's a bug when no hosts are available that match the criteria - the AJAX passes "preview_host_id"=>"null" and then you get a large 404 error in the page

Thanks, fixed.

domcleal commented:
> I was hoping for this to be a candidate for "Small and low risk ones may even make it through to 1.9." stated in branch out announcement

That was between 1.9 branching and RC1, we're weeks later now.

> 1.9.1 would be probably too late for remote execution

It's a plugin that hasn't even been released, I fail to see why it matters.

Changes right now for 1.9 are generally going to 1.9.1 as 1.9.0 stabilises for release.  It will give us longer to check for regressions.

ares commented:
> It's a plugin that hasn't even been released, I fail to see why it matters.

Our first version was meant to work with 1.9 we'd have more feedback from users running stable versions before second release. After reviewing dates, it would be OK if this could get in 1.9.1. Definitely much better than 1.10, the gap between our targeted release would be a week (based on 1.9.1 schedule).

domcleal commented:
Merged as de9e7ad for 1.9.1, thanks @ares.

stbenjam commented:
This is a plugin API change, which moves methods like kickstart_attributes from the unattended controller to the Foreman renderer, it should not make 1.9.1.

http://projects.theforeman.org/issues/11195

domcleal commented:
> This is a plugin API change, which moves methods like kickstart_attributes from the unattended controller to the Foreman renderer, it should not make 1.9.1.

Yeah, not really a published API, but I agree and would move it to 1.10.0.  Since it's been requested for 1.9 above, if you have a fix (it appears not, the ticket's still New), then we could instead bring it forward to 1.9.0-RC3 rather than push it back.

For now, I'll move to 1.10.0 unless you can confirm that you'll have a fix in time for RC3.  Cheers.

stbenjam commented:
Very little is a published API for plugins, but I think it's fair that public method signatures are not changed in a patch release -- or at least that should be as rare as possible.

It's easy to fix, but including it in Katello 2.3.0 GA means it won't work with 1.9.0 GA which will have the old code.

We could synchronize 1.9.1 and 2.3.1 to go at the same time, if this is really that important.

stbenjam commented:
(...although these are protected methods, so my argument isn't a particularly good one, but there's not really any good way to override the URL on a post-host basis.  We should probably contribute some code to Foreman to make the installation URL a host method that a plugin could change.  Overriding kickstart_attributes is hacky)

domcleal commented:
> Very little is a published API for plugins, but I think it's fair that public method signatures are not changed in a patch release -- or at least that should be as rare as possible.

Yeah, indeed, which is why I put patches containing refactorings on the next major release by default.  It was only because @ares asked so politely.

> It's easy to fix, but including it in Katello 2.3.0 GA means it won't work with 1.9.0 GA which will have the old code.

If you can fix it in time for GA, then I can include this in 1.9.0-RC3 - will that be OK?

stbenjam commented:
> If you can fix it in time for GA, then I can include this in 1.9.0-RC3 - will that be OK?

Yea that's fine - although it might be trickier than I thought.

I just tested it, and the failure isn't actually that the methods moved, it's the addition of this `if @host.medium`, it looks Hosts with a content view get the media box greyed out so we control it.  No idea why we do that.

New:
https://github.com/theforeman/foreman/commit/de9e7adaa579db01d043d6a04a83c946a3d33850#diff-6632a0077c474a44423343eb597f9d72R162

Original:
https://github.com/theforeman/foreman/commit/de9e7adaa579db01d043d6a04a83c946a3d33850#diff-1bab949bd7fbcda429d6b1dfe0b41b41L229

stbenjam commented:
@ares Can you help me understand why the if condition was added? This essentially makes medium required on a Host for provisioning, where it wasn't before. 

I can fix Katello, but it requires extending Foreman::Renderer which I seem to be having trouble doing.  If the condition could be removed, it would make this much easier.

Otherwise I'm not confident I'll have time to make Katello 2.3 compatible with this change.

ares commented:
@stbenjam see https://github.com/theforeman/foreman/pull/2524#discussion_r34249308, the issue was that when you tried to render template for host that was provisioned from image it didn't have medium and therefore it failed. I think we shouldn't try to set @mediapath if we don't have installation medium... 

as noted in the issue, @jlsherrill asked me about this few days ago and I think he was working on the fix, maybe he has something? IIRC we concluded that best would be to wrap `kickstart_attributes` method and make sure you call the original method and at the end fill in @mediapath if needed (based on hosts provisioning method)

stbenjam commented:
@ares That comment only talks about ptable.  Why should you be able on an image-rendered host be able to preview a provision template? That doesn't make sense.

Yes, I realize its probably fixable from Katello's side, but I'm not exactly sure what I should do other than just revert your change inside of Katello.  @jlsherrill Is there some better check I should do? `if @host.content_source`?

jlsherrill commented:
@stbenjam I would first check for an associated media, if there is none then just call what is currently Redhat#medium_uri_with_content_uri (but should probably be called something else as alias_method_chain wouldn't be needed.

ares commented:
@stbenjam the comment applies for both media and ptable as both received the same fix, sorry I should have been more clear. The reason was that when you're in the host detail you have no idea how was the host provisioned but you can see provisioning templates, since the OS for the host has association. When you clicked review on this template you got 500. Maybe it's a different error that we allow users to review provisioning templates in this case but I still think we shouldn't fail rendering if we're missing media for whatever reason, e.g. user tries to preview the result before he selected the medium.

stbenjam commented:
Alright fine, @domcleal this can make 1.9.0 RC3, if you can.  I'll release Katello 2.3 RC2 concurrently tomorrow, and try to get the fix PR opened today.

domcleal commented:
Thanks!
